<a href="https://colab.research.google.com/github/jppeirce/DSC210-Foundations-of-Data-Science/blob/main/Homework/HW01-foundations_and_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Homework 1: Foundations of Data Science and Data Types

**DSC 210 Foundations of Data Science**

*Covers Module 1 (What is Data Science?) and Module 2 (What Is Data? Big Data and the Data Science Workflow).*

References:
- Module 1 and Module 2 course notebooks
- [Hands-on Introduction to Data Science with Python](https://florian-huber.github.io/data_science_course/) (CC BY-NC-SA 4.0)
- `titanic` and `diamonds` datasets, included with the `seaborn` package

## Key Concepts

- Apply a working definition of data science and the three-circle framing of the field
- Use the 5-C checklist to evaluate the ethical dimensions of a data science project
- Identify cases, features, and metadata in a structured dataset
- Classify a feature by measurement scale and decide which summaries are meaningful
- Judge whether a dataset is the *right* data for a question, and locate work on the
>ASK -> GET -> EXPLORE -> MODEL -> COMMUNICATE loop

## Instructions

**Total: 50 points.** Point values are listed with each problem.

- Answer written questions in the markdown cells marked **Your answer:**.
    - Double-click the cell to edit it.
- Run every code cell. You are not asked to write code from scratch on this assignment. Where a problem asks you to change code, change only what the problem names and leave the rest alone.
- Several problems ask you to **predict before you run**. Write the prediction down first.
- Submit the completed `.ipynb` file, with all cells run, to Canvas.

---
## Part 0. Setup check (2 points)
---

Run the cell below. It loads the two packages and the dataset for this assignment. If it runs without an error and prints a shape, the Colab environment is working.

If you get an error, do not spend an hour on it alone. Bring the error message to office hours ask me before/after class.

In [ ]:
import seaborn as sns
import pandas as pd

titanic = sns.load_dataset('titanic')

print('pandas version :', pd.__version__)
print('seaborn version:', sns.__version__)
print('titanic shape  :', titanic.shape)

pandas version : 2.2.3
seaborn version: 0.13.2
titanic shape  : (891, 15)


**Problem 0.** (2 points) Did the cell above run without an error? If yes, say so and report the shape it printed. If no, paste the error message.

**Your answer:** Yes the cell ran as intended



---
## Part A. Short answer (15 points)
---
Answer in complete sentences. Two or three sentences is plenty for each part.

### Problem A1. Is this data science? (3 points)

In Module 1 we settled on a working definition:

> *Data science is the art (or practice) of gaining, interpreting, and communicating insights from complex data through digital techniques.*

We also noticed that the tasks people call "definitely data science" tend to involve **learning from data** (not just computing a statistic), **scale**, and **a decision or insight as the output**.

For each task below, decide whether it is best described as data science, statistics, software engineering, or some combination. Justify each answer in one or two sentences using at least one of the three features above.

| # | Task |
| --- | --- |
| a |A hospital builds a system that flags which discharged patients are most likely to be readmitted within 30 days, using ten years of patient records. |
| b |  In 2020 Pfizer-BioNTech did a clinical trial involving 43,000 people for the COVID-19 mRNA vaccine in which they compared the proportion of people who got sick in a vaccinated group versus an unvaccinated group. |
| c | A student writes a script that downloads every UWL course description and counts how many contain the word "data." |

**Your answer:**

(a) This is an example of data science. This is data science because it uses a large scale (10 years of data) that cannot be computed by hand and it learns from the data and makes a decision on readmittance.

(b) This is an example of statistics because it works on a smaller scale and statistically compares proprotions of two groups. This does not work on a large scale to learn from data and make a decision like data science does.

(c) This is an example of software engineering because a program works to count occurances of a word. This does not interpret the data, it only fetches data that matches the conditons defined.


### Problem A2. Where is the gap? (3 points)

Recall the three circles from Module 1: **Math & Statistics**, **Hacking Skills**, and **Substantive Expertise**.

A university's advising center is building a system that flags students who look "disengaged" in a course, so their instructor can reach out. The system uses only what Canvas records: how many times a student logs in, how many minutes each page stays open, and how early assignments are submitted.

The team is three people: two software engineers who built the data pipeline, and one statistician who fit the model. Nobody on the team has taught one of these courses, and nobody has asked a student how they actually use a Canvas course site.

(a) Which circle is the team missing?

(b) Describe one *specific* way the missing circle could cause this project to fail or to do harm. Be concrete. "They might not understand the data" is not specific enough.

**Your answer:**

(a) The team is missing the substantive expertise circle. This circle deals with someone who has practical experience with the data.

(b) Without a person with substantive expertise, this project could fail because they would not understand the reasons for someone to submit an asignment early. Only someone with substantive expertiese, like a teacher or student who uses canvas, would know that there may be other reasons for a student to submit an assinment early rather than engagement. This oversight can be detremental to the success of the project.


### Problem A3. Measurement scales (6 points)

For each feature, give (i) the measurement scale (nominal, ordinal, interval, or ratio) and (ii) one summary or operation that *is* meaningful/meaningless for it.

| # | Feature |
| --- | --- |
| a | The jersey number of a basketball player |
| b | The year a song was released |
| c | A t-shirt size (S, M, L, XL) |
| d | The body mass of a penguin, in grams |
| e | A UWL student's home zip code |
| f | The temperature of a lab freezer, in degrees Celsius |

**Your answer:**

| # | Scale | A meaningful summary or operation |
| --- | --- | --- |
| a | nominal | classification |
| b | interval | mean |
| c | ordinal | median |
| d | ratio | standard deviation |
| e | nominal | classification |
| f | interval | subtraction |


### Problem A4. Structured, unstructured, and metadata (3 points)

UWL Dining Services has three collections of material:

- (i) A short text file recording who set up the swipe system, the date it went live, and a note that the card readers at one location were offline for three days in October.
- (ii) Roughly 4,000 photos that students posted publicly with the dining hall tagged as the location.
- (iii) A table with one row per meal swipe, recording the student ID, the dining location, the date, and the time.

(a) Classify each of (i), (ii), and (iii) as structured data, unstructured data, or metadata.

(b) For the structured one, identify the **cases** and list the **features**.

(c) Explain in one sentence why (i) matters to anyone analyzing (iii).

**Your answer:**

(a) (i) is an example of metadata. (ii) is an example of unstructured data. (iii) is an example of sturctured data

(b) The cases for this example would be each meal swipe. The features for this example would be the student ID, the dining location, the date, and the time.

(c) The metadata is important to the structured data because it gives context on what the data is and why it was taken, and can help those analysing to interpret the meaning of the data.


---
## Part B. Reading and predicting code (8 points)
---

The RMS Titanic was a massive British passenger ship carrying 2,240 passengers and crew on its first voyage. On April 15, 1912, the ship sank after striking a large iceberg, resulting in the deaths of more than 1,500 people in the cold waters of the North Atlantic.

The Titanic dataset `titanic` in seaborn contains information about passengers aboard the RMS Titanic, including their survival status, age, sex, passenger class, fare, and family relationships.

### Problem B1. When a number is not a number (4 points)

The cell below computes the mean of two features:
- `fare` measures the amount the passenger payed for their ticket
- `pclass` records which passenger class a person traveled in, stored as the number 1 = first-class, 2 = second-class, or 3 = third-class.

Both lines run without an error.

Before running the code, predict whether both of these outputs be *meaningful*?

**Your prediction:** Only the fare will be meaningful, because a mean does not apply to an ordinal feature.



In [ ]:
print('mean fare  :', round(titanic['fare'].mean(), 2))
print('mean pclass:', round(titanic['pclass'].mean(), 2))

mean fare  : 32.2
mean pclass: 2.31


(a) Python computed a mean for `pclass` without complaining. Explain why that number is not a meaningful summary. Name the measurement scale of `pclass` in your answer.

(b) Compare this to Module 2, where asking for the mean of `species` produced a `TypeError`. Why did `pclass` fail *silently* when `species` failed *loudly*? Which of the two is more dangerous for an analyst, and why?

(c) In the cell below, change **only the method name** `.mean()` so that the line produces an appropriate summary of `pclass`. Everything else stays as it is. Then run it.

**Your answer:**

(a) This number is not a meaningful summary because pclass is an ordinal scale. This scale is not measured in numerical values but can still be stored as them, in this case the numerical values are merely categorical, thus, a mean is insignifcant.

(b) This does not compute a type error because the value is still numerical. The system can still compute this value, but the value is does not have significance, this is still an error. The silent error is much more dangerous for an analyst because the system does not warn the analyst directly.


In [ ]:
# Problem B2(c): change only .mean() to an appropriate method for pclass
print(titanic['pclass'].median())

3.0


### Problem B2. What is missing, and why (4 points)

In python, some of the data structures we will come with special tools called methods (we will learn more about these in Module 3). Consider the following methods:

*   `.sum()` calculates the total sum of the values
*   `isna()` detects missing values and returns a boolean object (True indicates a missing value and False indicates a valid value).

The cell below counts how many values are missing in each feature.

In [ ]:
print(titanic.isna().sum())

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


(a) A classmate says: "A blank in the `deck` column means that passenger did not have a deck." Explain why that is almost certainly wrong, and give a more plausible explanation for the blanks.

(b) Module 2 makes the claim that a dataset is never just "found." Use these missing values to explain what that claim means, in one or two sentences.

**Your answer:**

(a) The blank in the deck meaning that the passenger has no deck is almost certianly is wrong because practically each passenger on the titanic has a deck. This was most likely not able to be recorded at the time and the passenger was still recorded.

(b) When recording data, the process is very intentional. When collecting, someone may attempt to collect something, but missing values may happen naturally as part of the process.


---
## Part C. Guided problems (10 points)
---

### Problem C1. A feature audit (4 points)

Run the cell below. It shows how Python is storing each feature (the `dtype`) along with the first few cases.

In [ ]:
print(titanic.dtypes)
print()
titanic[['survived', 'pclass', 'sex', 'age', 'fare']].head()

survived          int64
pclass            int64
sex              object
age             float64
sibsp             int64
parch             int64
fare            float64
embarked         object
class          category
who              object
adult_male         bool
deck           category
embark_town      object
alive            object
alone              bool
dtype: object



,survived,pclass,sex,age,fare
0,0,3,male,22.0,7.2500
1,1,1,female,38.0,71.2833
2,1,3,female,26.0,7.9250
3,1,1,female,35.0,53.1000
4,0,3,male,35.0,8.0500


(5 points) Complete the table below for the five features shown. For the fourth column, answer yes or no: does the way Python stores the feature match its measurement scale?

| Feature | dtype (as stored) | Measurement scale | dtype matches scale? | One meaningful summary |
| --- | --- | --- | --- | --- |
| `survived` | int64 | ordinal | no | This is stored as a numerical 0 or 1 for yes or no category. |
| `pclass` | int64 | ordinal | no | This is stored as a numerical 1, 2, or 3 for each class category. |
| `sex` | object | ordinal | yes | This is stored as an object with the category name for each sex. |
| `age` | float64 | ratio | yes | This is stored as a numerical float for each age. |
| `fare` | float64 | ratio | yes | This is stored as a numerical float for each fare price. |



### Problem C2. Ordinal scales, and what makes data big (6 points)

Run the cell below. It loads a second dataset, `diamonds`, and shows the categories of two of its features.

In [ ]:
diamonds = sns.load_dataset('diamonds')
print('diamonds shape:', diamonds.shape)
print()
print('cut categories    :', list(diamonds['cut'].unique()))
print('clarity categories:', list(diamonds['clarity'].unique()))

539400
diamonds shape: (53940, 10)

cut categories    : ['Ideal', 'Premium', 'Good', 'Very Good', 'Fair']
clarity categories: ['SI2', 'SI1', 'VS1', 'VS2', 'VVS2', 'VVS1', 'I1', 'IF']


In the diamond trade, `cut` runs from Fair (worst) to Ideal (best), and `clarity` runs from I1 (most inclusions) to IF (internally flawless).

(a) (2 points) Explain why `cut` is **ordinal** and not **nominal**.

(b) (2 points) Explain why `cut` is **ordinal** and not **interval**. Your answer should say something about the *distance* between categories.

(c) (2 points) The `diamonds` dataset has far more rows than `penguins` (344) or `titanic` (891). Using the V's from Module 2, Section 4, decide whether `diamonds` counts as **big data**. Refer to at least two of the V's, and say what specifically would have to change for your answer to flip.

**Your answer:**

(a) Cut is an ordianal scale rather than nominal because it has ordered categories. These categories range from ideal, premium, good, very good, and fair and act on an ordered scale, rather than an unordered nominal scale

(b) Cut is ordinal rather than interval because the distace between the categories are unequal. With interval data each interval is evenly measured making cut a better fit for ordinal scale.

(c) The data would not be considered to be big. This is because the data lacks variety. Each case describes a single diamond and its qualities; there is not multiple various structures of data incoming. This is also because the data lacks velocity. The data is not contiously generating; it remains the same size. To change this to big dat, data needs to be continously generating at a large volume and have various types of data flowing in.


---
## Part D. Application (12 points)
---

### Problem D1. An inventory of your own data (7 points)

In class we brainstormed what data is collected about you. Now do it carefully and in writing.

(a) (3 points) List **eight** distinct data streams that recorded something about you in the last 24 hours. Include apps and services, physical sensors, and institutional records. Complete the table.

| # | Data stream | What is being observed or measured | Who stores it | Could this have been collected in 1996? |
| --- | --- | --- | --- | --- |
| 1 | Youtube | My video activiy | Google | No |
| 2 | Banking information | My transaction history | Altra Credit Union | Yes |
| 3 | Grocery store | My Purchases | The Store and its corperate company | Yes |
| 4 | TV (Roku) | My browsing data | Roku | No |
| 5 | Canvas | My activity and scores | Canvas and UWL | No |
| 6 | Google Colab | My activity within this document | Google and GitHub | No |
| 7 | Gmail | My emails, email content, and history | Google | No |
| 8 | Google | My search activity | Google | No |

(b) (3 points) Choose **one** stream from your table and apply the 5-C checklist. One sentence of assessment and a risk level (Low / Medium / High) for each criterion.

**Stream of Choice:** Youtube
| Criterion | Your assessment | Risk |
| --- | --- | --- |
| Consent | By using the app you consent to data collection and useage | Medium |
| Clarity | The app uses data for algorithmically picking videos for you | Medium |
| Consistency | The app consistently collects the data from any interaction | Low |
| Control and Transparency | The app is fairly transparent with the data it collects and it is eviedinced by the algorithm. | Medium |
| Consequences and Harm | The data is saved amongst all google services and general history on the account in a non harmful way. | Low |

(c) (1 point) For that same stream, name one thing that could be inferred about you that you never directly shared. Explain how the inference would work.

**Your answer:**

(a) *(fill in the table above)*

(b) *(fill in the table above)*

(c) Collecting the data youtube can determine your intrests and cater their recomandations for advertisements and videos. They can infer your intrests by understanding how you interact with the videos that you do and don't watch


### Problem D2. Case study: are the bike lanes working? (8 points)

A city council has the following *objective*:

>  Determine whether the city's bike lanes are making cycling safer.

Currently this is a statement that does not provide a question you can answer with data. Walk it onto the course loop.

```
ASK  ->  GET  ->  EXPLORE  ->  MODEL  ->  COMMUNICATE
 ^                                            |
 |________ findings raise new questions ______|
```

(a) (2 point) **ASK.** Turn the statement above into one specific, answerable question. Two quick checks:

- Could data settle it?

- Is it clear what *kind of number* would count as an answer?

(b) (2 points) **GET.** Name two data sources you would pursue, using the table of sources in Module 2, Section 5. One should be data that already exists; one should be data you would create yourself. For each, name the source type and state its catch.

| Source | Type (from Section 5) | The catch |
| --- | --- | --- |
| Bike realated accidents history | Public dataset | It my not be current, or the best fit for my question |
| Collect bike lane incident traffic data | Create your own | I now own its design, quality, and ethics |

(c) (3 points) **Garbage in, garbage out.** The council already has one thing: four years of police collision reports. Every collision below was reported to police, and every number is accurate.

| Street type | Cyclist collisions, 2019 | Cyclist collisions, 2023 |
| --- | --- | --- |
| Streets that received protected bike lanes | 42 | 61 |
| Comparable streets, no lanes added | 38 | 40 |

A council member reads this table and concludes: *"Collisions on the lane streets jumped 45% while barely moving elsewhere. The lanes made cycling more dangerous. We should stop building them."*

Is she right?

Answer in two parts. First, state whether the table is sufficient to support her conclusion. Second, if it is not, name **specifically** what else you would need to know before anyone could interpret these counts, and explain what that missing information would change. Be concrete: name the quantity, not just "more context."

(d) (1 point) **EXPLORE, MODEL, and the loop.** Fill in one row of the data-information-knowledge progression for this project. Then, in one sentence below the table, state a new question that acting on your finding would raise.

| Stage | For my bike lane project |
| --- | --- |
| Data | Collect raw data on details about each collision that happens after putting a new bike lane on a road |
| Information |  |
| Knowledge |  |

**Your answer:**

(a) Do the city's bike lanes reduce bike realated trafic accidents on roads that use them?

(b) *(fill in the table above)*

(c) The data provided is not sufficent to support her conclusion. To have sufficent data she would need to know how often people cycled in those two years listed, as well as how often people drove on those roads in those two years.

(d) *(fill in the table above, then the new question here)*
I chose the data row for this option. The data I collect would raise the question of: *"how frequently did people drive or bike on these roads?"*
